### ISSUE TITLE
Cluster-level resources

### ISSUE THREAD

**Issue by jcrist** [dask/distributed#6382]
Currently we have worker-level resources, which are useful for limiting access to things that are limited per-worker (memory, GPU, etc...). Sometimes tasks will consume resources that are globally limited though (database connections for example). Right now users have a few options:

- Divide the global resource evenly between workers and set per-worker resources accordingly. This doesn't play well with adaptive scaling, and requires knowing the cluster size to apply the limit.
- Use a `distributed.Semaphore` to limit the active usage of a global resource. This effectively sets a global limit, but doesn't let the scheduler see the concurrency limit. As such, the schedule may allocate more resource-consuming tasks than can currently run, with the excess tasks all taking up an execution thread while waiting on the semaphore to run.

Cluster level resources would let the scheduler take this limit into consideration during scheduling.

**Comment by mrocklin** [comment]
Do we know cases where the semaphore approach wouldn't be good enough?

**Comment by gjoseph92** [comment]
Because there's no backpressure on queuing tasks on workers, I'm not sure what this would add on the task-scheduling side. The scheduler will send every task whose dependencies are ready to a worker, regardless of how saturated that worker is. If you have some "normal" tasks to run, and some semaphore-bound tasks to run, the scheduler will submit all of them to workers as long as they're runnable.

> with the excess tasks all taking up an execution thread while waiting on the semaphore to run

Could `secede`/`rejoin` be used here?

The point being, because the scheduler never holds back from scheduling runnable tasks, the problem of semaphore-bound tasks gumming up the cluster isn't a scheduler problem. It's just about workers not recognizing that they could run some of their other ready tasks while waiting for the semaphore to become available.

Interestingly, if cluster-level resources existed, that would be an easy way to address https://github.com/dask/distributed/issues/6360.

**Comment by fjetter** [comment]
This is a duplicate (sort of) of https://github.com/dask/distributed/issues/4104 where a user requested to have semaphores on task level such that the scheduler only schedules as many tasks as a semaphore would allow. Back then we didn't have much infrastructure to support this (e.g. task annotations were just being designed/implemented). I think we're in a much better position right now to implement this and I can see value in it.

I think a robust secede/rejoin logic for the semaphore would solve this problem as well but there are a few subtle race conditions possible that I'm not sure we can resolve in a robust fashion. I mentioned the problem of resource starvation already on the original ticket, see https://github.com/dask/distributed/issues/4104#issuecomment-700731853
in a nutshell; assume there are N total leases / semaphore value. >N tasks are scheduled on one worker. allowed threads per worker M<<N. There is a race condition upon rejoining that could cause resource starvation. A seceded task is only allowed to rejoin if it can acquire a semaphore lease. However, it is also only allowed to rejoin if there is space on the threadpool.
There is also a race condition about priorities of rejoining tasks vs ordinary tasks which we currently don't handle which could starve our system. There are likely more things that could go wrong or are at least unpleasant side effects.

We could likely work around this by using some kind of optimistic concurrency control (timeout on rejoin, release semaphore on failure, try again, maybe use a lock to avoid scheduling new tasks if an existing task wants to rejoin, etc.)

----

I do like the approach of cluster wide resources much more. There are surely complications as well but I think we should be able to mostly replicate the state machine logic from the worker to the scheduler, i.e. introduce a "sibling state" `constrained` which works almost as the `waiting` state. We perform mostly the same transitions for constrained as for waiting. The worker likely shouldn't even care about this. From its perspective this is just an ordinary task.

**Comment by dask-user** [comment]
I had started a topic on the discourse forum that seems to be an exact duplicate of this issue (including providing the same 2 alternative approaches). Would like to link here to provide support/additional context on usefulness of this feature.

https://dask.discourse.group/t/global-resources-similar-to-worker-resources/1579/3?u=dask-user
